File: sdks/go/providers/inworld/protocol.go
package inworld

import (
	"bytes"
	"encoding/base64"
	"encoding/binary"
	"encoding/json"
	"errors"
	"fmt"
	"math"
	"strings"

	out "github.com/speechswitch/client/sdks/go/generated/inworld_output"
	"github.com/speechswitch/client/sdks/go/runtime"
)

func decode(data []byte) (map[string]any, error) {
	if !runtime.ValidUTF8JSON(data) {
		return nil, errors.New("Inworld returned invalid JSON")
	}
	var value map[string]any
	if json.Unmarshal(data, &value) != nil || value == nil {
		return nil, errors.New("Inworld returned an invalid object")
	}
	return value, nil
}
func status(value map[string]any) error {
	if value == nil {
		return errors.New("Inworld returned an invalid object")
	}
	code, ok := value["code"].(float64)
	if _, present := value["code"]; present && (!ok || math.Trunc(code) != code || math.Abs(code) > 9007199254740991) {
		return errors.New("Inworld returned an invalid status code")
	}
	message, mOK := value["message"].(string)
	if _, present := value["message"]; present && !mOK {
		return errors.New("Inworld returned an invalid status message")
	}
	if ok && code != 0 {
		if !mOK {
			message = fmt.Sprintf("Inworld status %d", int64(code))
		}
		return &Error{Message: message, Code: runtime.Some(int64(code))}
	}
	return nil
}
func result(data []byte) (map[string]any, error) {
	packet, err := decode(data)
	if err != nil {
		return nil, err
	}
	if value, present := packet["error"]; present {
		fields, _ := value.(map[string]any)
		if err := status(fields); err != nil {
			return nil, err
		}
		message, ok := fields["message"].(string)
		if !ok {
			message = "Inworld synthesis failed"
		}
		return nil, &Error{Message: message}
	}
	fields, ok := packet["result"].(map[string]any)
	if !ok {
		return nil, errors.New("Inworld returned an invalid object")
	}
	return fields, nil
}
func seconds(value any) (float64, error) {
	v, ok := value.(float64)
	if !ok || v < 0 || math.IsNaN(v) || math.IsInf(v*1000, 0) {
		return 0, errors.New("Inworld returned an invalid timestamp")
	}
	return v * 1000, nil
}
func timestamps(value any) ([]out.InworldTimestamp, error) {
	info, ok := value.(map[string]any)
	if !ok {
		return nil, errors.New("Inworld returned an invalid object")
	}
	marks := []out.InworldTimestamp{}
	for _, kind := range []string{"word", "character"} {
		raw, present := info[kind+"Alignment"]
		if !present {
			continue
		}
		alignment, ok := raw.(map[string]any)
		if !ok {
			return nil, errors.New("Inworld returned an invalid object")
		}
		values, vOK := alignment[kind+"s"].([]any)
		starts, sOK := alignment[kind+"StartTimeSeconds"].([]any)
		ends, eOK := alignment[kind+"EndTimeSeconds"].([]any)
		if !vOK || !sOK || !eOK || len(values) != len(starts) || len(values) != len(ends) {
			return nil, errors.New("Inworld returned mismatched timestamp arrays")
		}
		for index, raw := range values {
			token, ok := raw.(string)
			if !ok {
				return nil, errors.New("Inworld returned an invalid alignment token")
			}
			start, err := seconds(starts[index])
			if err != nil {
				return nil, err
			}
			end, err := seconds(ends[index])
			if err != nil {
				return nil, err
			}
			if end < start {
				return nil, errors.New("Inworld returned a reversed timestamp range")
			}
			mark := out.InworldTimestamp{Value: token, StartTimeMs: start, EndTimeMs: runtime.Some(end), Kind: out.InworldTimestampKindAsCharacter{}}
			if kind == "word" {
				mark.Kind = out.InworldTimestampKindAsWord{}
				mark.WordIndex = runtime.Some(float64(index))
			}
			marks = append(marks, mark)
		}
		raw, present = alignment["phoneticDetails"]
		if kind != "word" || !present {
			continue
		}
		details, ok := raw.([]any)
		if !ok {
			return nil, errors.New("Inworld returned invalid phonetic details")
		}
		for _, raw := range details {
			detail, ok := raw.(map[string]any)
			if !ok {
				return nil, errors.New("Inworld returned an invalid object")
			}
			index, iOK := detail["wordIndex"].(float64)
			phones, pOK := detail["phones"].([]any)
			if !iOK || math.Trunc(index) != index || index < 0 || index >= float64(len(values)) || !pOK {
				return nil, errors.New("Inworld returned an invalid phonetic word index")
			}
			for _, raw := range phones {
				phone, ok := raw.(map[string]any)
				if !ok {
					return nil, errors.New("Inworld returned an invalid object")
				}
				symbol, sOK := phone["phoneSymbol"].(string)
				viseme, vOK := phone["visemeSymbol"].(string)
				_, vPresent := phone["visemeSymbol"]
				if !sOK || (vPresent && !vOK) {
					return nil, errors.New("Inworld returned an invalid phone or viseme")
				}
				start, err := seconds(phone["startTimeSeconds"])
				if err != nil {
					return nil, err
				}
				duration, err := seconds(phone["durationSeconds"])
				if err != nil {
					return nil, err
				}
				end := start + duration
				if math.IsInf(end, 0) {
					return nil, errors.New("Inworld returned an invalid timestamp")
				}
				mark := out.InworldTimestamp{Kind: out.InworldTimestampKindAsPhoneme{}, Value: symbol, StartTimeMs: start, EndTimeMs: runtime.Some(end), WordIndex: runtime.Some(index)}
				marks = append(marks, mark)
				if vOK {
					mark.Kind = out.InworldTimestampKindAsViseme{}
					mark.Value = viseme
					marks = append(marks, mark)
				}
			}
		}
	}
	return marks, nil
}

func audioOutput(packet map[string]any, timed, chunk bool, id runtime.Optional[string], wave *waveStream) (out.SynthesisItem, error) {
	if value, present := packet["status"]; present {
		fields, _ := value.(map[string]any)
		if err := status(fields); err != nil {
			return nil, err
		}
	}
	rawAudio, aPresent := packet["audioContent"]
	rawMarks, mPresent := packet["timestampInfo"]
	_, usage := packet["usage"]
	if !aPresent && !mPresent && !usage {
		return nil, errors.New("Inworld returned no audio or alignment")
	}
	var audio []byte
	var err error
	if aPresent {
		encoded, ok := rawAudio.(string)
		if !ok {
			return nil, errors.New("Inworld returned invalid audio content")
		}
		audio, err = base64.StdEncoding.DecodeString(encoded)
		if err != nil || strings.ContainsAny(encoded, "\r\n") {
			return nil, errors.New("Inworld returned invalid base64 audio")
		}
		if wave != nil {
			audio, err = wave.audio(audio)
			if err != nil {
				return nil, err
			}
		}
	}
	marks := []out.InworldTimestamp{}
	if mPresent {
		marks, err = timestamps(rawMarks)
		if err != nil {
			return nil, err
		}
	}
	if !timed {
		if len(audio) == 0 {
			return nil, nil
		}
		return out.SynthesisItemAsBytes{Value: audio}, nil
	}
	if !aPresent && !mPresent {
		return nil, nil
	}
	if chunk {
		if !aPresent {
			return nil, errors.New("Inworld omitted audio from synchronized alignment")
		}
		return out.SynthesisItemAsChunk{Value: out.InworldChunkEnvelope{Audio: audio, Timestamps: marks, CorrelationId: id}}, nil
	}
	envelope := out.InworldTimelineEnvelope{Timestamps: marks, CorrelationId: id}
	if aPresent {
		envelope.Audio = runtime.Some(audio)
	}
	return out.SynthesisItemAsTimeline{Value: envelope}, nil
}

// Only the first native flush retains an open-ended PCM WAV header.
type waveStream struct {
	pending, format []byte
	header          bool
}

func (w *waveStream) boundary() error {
	if len(w.pending) > 0 {
		return errors.New("Inworld returned an incomplete WAV header")
	}
	w.header = true
	return nil
}
func (w *waveStream) audio(value []byte) ([]byte, error) {
	if !w.header {
		return value, nil
	}
	w.pending = append(w.pending, value...)
	data := w.pending
	if len(data) < 12 {
		return nil, nil
	}
	if string(data[:4]) != "RIFF" || string(data[8:12]) != "WAVE" {
		return nil, errors.New("Inworld returned an invalid WAV header")
	}
	var format []byte
	for offset := 12; offset+8 <= len(data); {
		size := uint64(binary.LittleEndian.Uint32(data[offset+4 : offset+8]))
		if string(data[offset:offset+4]) == "data" {
			if format == nil {
				return nil, errors.New("Inworld WAV omitted its format")
			}
			first := w.format == nil
			if !first && !bytes.Equal(w.format, format) {
				return nil, errors.New("Inworld changed WAV format between flushes")
			}
			w.format, w.header, w.pending = bytes.Clone(format), false, nil
			if !first {
				return data[offset+8:], nil
			}
			binary.LittleEndian.PutUint32(data[4:8], 0xffffffff)
			binary.LittleEndian.PutUint32(data[offset+4:offset+8], 0xffffffff)
			return data, nil
		}
		if uint64(offset)+8+size > 1048576 {
			return nil, errors.New("Inworld WAV header is too large")
		}
		if offset+8+int(size) > len(data) {
			return nil, nil
		}
		if string(data[offset:offset+4]) == "fmt " {
			if size < 16 || binary.LittleEndian.Uint16(data[offset+8:offset+10]) != 1 || binary.LittleEndian.Uint16(data[offset+22:offset+24]) != 16 {
				return nil, errors.New("Inworld returned an invalid PCM WAV format")
			}
			channels := uint32(binary.LittleEndian.Uint16(data[offset+10 : offset+12]))
			if channels == 0 || uint32(binary.LittleEndian.Uint16(data[offset+20:offset+22])) != channels*2 {
				return nil, errors.New("Inworld returned an invalid PCM WAV format")
			}
			format = data[offset+8 : offset+8+int(size)]
		}
		offset += 8 + int(size) + int(size%2)
	}
	return nil, nil
}

File: sdks/rust/src/providers/inworld/protocol.rs
use crate::{base64, generated::inworld_output::*, http::TransportError, json::Raw};
use std::collections::BTreeMap;

#[derive(Debug, PartialEq, Eq)]
pub struct Error {
    pub message: String,
    pub status: Option<u16>,
    pub code: Option<i64>,
}
impl std::fmt::Display for Error {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str(&self.message)
    }
}
impl std::error::Error for Error {}
pub(super) fn failure(message: &str) -> TransportError {
    Box::new(std::io::Error::new(
        std::io::ErrorKind::InvalidData,
        message,
    ))
}
type Fields<'a> = BTreeMap<String, Raw<'a>>;
pub(super) fn decode(data: &str) -> Result<Fields<'_>, TransportError> {
    Raw::parse_exact(data)
        .map_err(|_| failure("Inworld returned invalid JSON"))?
        .object()
        .map_err(|_| failure("Inworld returned an invalid object"))
}
pub(super) fn status(fields: &Fields<'_>) -> Result<(), TransportError> {
    let code = fields
        .get("code")
        .map(|v| {
            v.number()
                .map_err(|_| failure("Inworld returned an invalid status code"))
        })
        .transpose()?;
    if code.is_some_and(|v| !v.is_finite() || v.fract() != 0.0 || v.abs() > 9007199254740991.0) {
        return Err(failure("Inworld returned an invalid status code"));
    }
    let message = fields
        .get("message")
        .map(|v| {
            v.string()
                .map_err(|_| failure("Inworld returned an invalid status message"))
        })
        .transpose()?;
    if let Some(code) = code.filter(|v| *v != 0.0) {
        return Err(Box::new(Error {
            message: message.unwrap_or_else(|| format!("Inworld status {}", code as i64)),
            status: None,
            code: Some(code as i64),
        }));
    }
    Ok(())
}
pub(super) fn result(data: &str) -> Result<Fields<'_>, TransportError> {
    let fields = decode(data)?;
    if let Some(error) = fields.get("error") {
        let error = error
            .object()
            .map_err(|_| failure("Inworld returned an invalid object"))?;
        status(&error)?;
        return Err(Box::new(Error {
            message: error
                .get("message")
                .and_then(|v| v.string().ok())
                .unwrap_or_else(|| "Inworld synthesis failed".into()),
            status: None,
            code: None,
        }));
    }
    fields
        .get("result")
        .and_then(|v| v.object().ok())
        .ok_or_else(|| failure("Inworld returned an invalid object"))
}
fn seconds(value: Option<&Raw<'_>>) -> Result<f64, TransportError> {
    let value = value
        .and_then(|v| v.number().ok())
        .filter(|v| *v >= 0.0 && (v * 1000.0).is_finite())
        .ok_or_else(|| failure("Inworld returned an invalid timestamp"))?;
    Ok(value * 1000.0)
}
fn timestamps(raw: Raw<'_>) -> Result<Vec<InworldTimestamp>, TransportError> {
    let info = raw
        .object()
        .map_err(|_| failure("Inworld returned an invalid object"))?;
    let mut marks = Vec::new();
    for kind in ["word", "character"] {
        let Some(raw) = info.get(&format!("{kind}Alignment")) else {
            continue;
        };
        let a = raw
            .object()
            .map_err(|_| failure("Inworld returned an invalid object"))?;
        let invalid = || failure("Inworld returned mismatched timestamp arrays");
        let values = a
            .get(&format!("{kind}s"))
            .and_then(|v| v.array().ok())
            .ok_or_else(invalid)?;
        let starts = a
            .get(&format!("{kind}StartTimeSeconds"))
            .and_then(|v| v.array().ok())
            .ok_or_else(invalid)?;
        let ends = a
            .get(&format!("{kind}EndTimeSeconds"))
            .and_then(|v| v.array().ok())
            .ok_or_else(invalid)?;
        if values.len() != starts.len() || values.len() != ends.len() {
            return Err(invalid());
        }
        for (index, raw) in values.iter().enumerate() {
            let value = raw
                .string()
                .map_err(|_| failure("Inworld returned an invalid alignment token"))?;
            let start = seconds(starts.get(index))?;
            let end = seconds(ends.get(index))?;
            if end < start {
                return Err(failure("Inworld returned a reversed timestamp range"));
            }
            marks.push(InworldTimestamp {
                kind: if kind == "word" {
                    InworldTimestampKind::Word(InworldTimestampKindWord)
                } else {
                    InworldTimestampKind::Character(InworldTimestampKindCharacter)
                },
                value,
                start_time_ms: start,
                end_time_ms: Some(end),
                source: None,
                word_index: (kind == "word").then_some(index as f64),
            });
        }
        let Some(raw) = a.get("phoneticDetails").filter(|_| kind == "word") else {
            continue;
        };
        for raw in raw
            .array()
            .map_err(|_| failure("Inworld returned invalid phonetic details"))?
        {
            let detail = raw
                .object()
                .map_err(|_| failure("Inworld returned an invalid object"))?;
            let invalid = || failure("Inworld returned an invalid phonetic word index");
            let index = detail
                .get("wordIndex")
                .and_then(|v| v.number().ok())
                .filter(|v| {
                    v.is_finite() && v.fract() == 0.0 && *v >= 0.0 && *v < (values.len() as f64)
                })
                .ok_or_else(invalid)?;
            let phones = detail
                .get("phones")
                .and_then(|v| v.array().ok())
                .ok_or_else(invalid)?;
            for raw in phones {
                let phone = raw
                    .object()
                    .map_err(|_| failure("Inworld returned an invalid object"))?;
                let invalid = || failure("Inworld returned an invalid phone or viseme");
                let symbol = phone
                    .get("phoneSymbol")
                    .and_then(|v| v.string().ok())
                    .ok_or_else(invalid)?;
                let viseme = phone
                    .get("visemeSymbol")
                    .map(|v| v.string().map_err(|_| invalid()))
                    .transpose()?;
                let start = seconds(phone.get("startTimeSeconds"))?;
                let end = start + seconds(phone.get("durationSeconds"))?;
                if !end.is_finite() {
                    return Err(failure("Inworld returned an invalid timestamp"));
                }
                marks.push(InworldTimestamp {
                    kind: InworldTimestampKind::Phoneme(InworldTimestampKindPhoneme),
                    value: symbol,
                    start_time_ms: start,
                    end_time_ms: Some(end),
                    source: None,
                    word_index: Some(index),
                });
                if let Some(value) = viseme {
                    marks.push(InworldTimestamp {
                        kind: InworldTimestampKind::Viseme(InworldTimestampKindViseme),
                        value,
                        start_time_ms: start,
                        end_time_ms: Some(end),
                        source: None,
                        word_index: Some(index),
                    });
                }
            }
        }
    }
    Ok(marks)
}
pub(super) fn audio(
    fields: &Fields<'_>,
    timed: bool,
    chunk: bool,
    id: Option<String>,
    wave: Option<&mut Wave>,
) -> Result<Option<SynthesisItem>, TransportError> {
    if let Some(raw) = fields.get("status") {
        status(
            &raw.object()
                .map_err(|_| failure("Inworld returned an invalid object"))?,
        )?;
    }
    if !["audioContent", "timestampInfo", "usage"]
        .iter()
        .any(|k| fields.contains_key(*k))
    {
        return Err(failure("Inworld returned no audio or alignment"));
    }
    let mut audio = fields
        .get("audioContent")
        .map(|v| {
            let encoded = v
                .string()
                .map_err(|_| failure("Inworld returned invalid audio content"))?;
            base64::decode(&encoded).ok_or_else(|| failure("Inworld returned invalid base64 audio"))
        })
        .transpose()?;
    if let (Some(wave), Some(bytes)) = (wave, audio.as_mut()) {
        *bytes = wave.audio(std::mem::take(bytes))?;
    }
    let marks = fields
        .get("timestampInfo")
        .map(|v| timestamps(*v))
        .transpose()?
        .unwrap_or_default();
    if !timed {
        return Ok(audio.filter(|v| !v.is_empty()).map(SynthesisItem::Bytes));
    }
    if audio.is_none() && !fields.contains_key("timestampInfo") {
        return Ok(None);
    }
    if chunk {
        return Ok(Some(SynthesisItem::Chunk(InworldChunkEnvelope {
            correlation: InworldChunkEnvelopeCorrelation,
            correlation_id: id,
            audio: audio
                .ok_or_else(|| failure("Inworld omitted audio from synchronized alignment"))?,
            timestamps: marks,
        })));
    }
    Ok(Some(SynthesisItem::Timeline(InworldTimelineEnvelope {
        correlation: InworldTimelineEnvelopeCorrelation,
        correlation_id: id,
        audio,
        timestamps: marks,
    })))
}

pub(super) struct Wave {
    pending: Vec<u8>,
    format: Option<Vec<u8>>,
    header: bool,
}
impl Wave {
    pub fn new() -> Self {
        Self {
            pending: Vec::new(),
            format: None,
            header: true,
        }
    }
    pub fn boundary(&mut self) -> Result<(), TransportError> {
        if !self.pending.is_empty() {
            return Err(failure("Inworld returned an incomplete WAV header"));
        }
        self.header = true;
        Ok(())
    }
    pub fn audio(&mut self, value: Vec<u8>) -> Result<Vec<u8>, TransportError> {
        if !self.header {
            return Ok(value);
        }
        self.pending.extend(value);
        let data = &mut self.pending;
        if data.len() < 12 {
            return Ok(Vec::new());
        }
        if &data[..4] != b"RIFF" || &data[8..12] != b"WAVE" {
            return Err(failure("Inworld returned an invalid WAV header"));
        }
        let mut offset = 12;
        let mut format = None;
        while offset + 8 <= data.len() {
            let size = u32::from_le_bytes(data[offset + 4..offset + 8].try_into().unwrap()) as u64;
            if &data[offset..offset + 4] == b"data" {
                let format: Vec<u8> =
                    format.ok_or_else(|| failure("Inworld WAV omitted its format"))?;
                let first = self.format.is_none();
                if self.format.as_ref().is_some_and(|v| *v != format) {
                    return Err(failure("Inworld changed WAV format between flushes"));
                }
                self.format = Some(format);
                self.header = false;
                let mut data = std::mem::take(data);
                if !first {
                    return Ok(data.split_off(offset + 8));
                }
                data[4..8].copy_from_slice(&u32::MAX.to_le_bytes());
                data[offset + 4..offset + 8].copy_from_slice(&u32::MAX.to_le_bytes());
                return Ok(data);
            }
            if offset as u64 + 8 + size > 1048576 {
                return Err(failure("Inworld WAV header is too large"));
            }
            let size = size as usize;
            if offset + 8 + size > data.len() {
                return Ok(Vec::new());
            }
            if &data[offset..offset + 4] == b"fmt " {
                if size < 16
                    || data[offset + 8..offset + 10] != [1, 0]
                    || data[offset + 22..offset + 24] != [16, 0]
                {
                    return Err(failure("Inworld returned an invalid PCM WAV format"));
                }
                let channels =
                    u16::from_le_bytes(data[offset + 10..offset + 12].try_into().unwrap()) as u32;
                let align =
                    u16::from_le_bytes(data[offset + 20..offset + 22].try_into().unwrap()) as u32;
                if channels == 0 || align != channels * 2 {
                    return Err(failure("Inworld returned an invalid PCM WAV format"));
                }
                format = Some(data[offset + 8..offset + 8 + size].to_vec());
            }
            offset += 8 + size + size % 2;
        }
        Ok(Vec::new())
    }
}
